### PR TITLE
Blog post for the Command provider GA

### PR DESCRIPTION
**Pull request by thomas11**
### Proposed changes

The [Command provider](https://www.pulumi.com/registry/packages/command/) is about to have its v1 released, marking GA.

### CODE REVIEW

File: content/blog/command-provider-ga/index.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
+---
+title: "Announcing the general availability of the Command provider"
+
+date: 2024-06-20T11:02:20+02:00
+
+draft: false
+
+meta_desc: >-
+    Version v1 of the Pulumi Command provider marks the provider’s general availability (GA).
+    The release also includes expanded support for copying assets.
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:10]: Slightly more active voice.
```suggestion
    The release expands support for copying assets between local and remote hosts.
```

File: content/blog/command-provider-ga/index.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+```yaml
+resources:
+  pet:
+    type: random:RandomPet
+
+  petFileCommand:
+    type: command:local:Command
+    properties:
+      create: !Sub 'touch "${pet.id}.txt"'
+      delete: !Sub 'rm "${pet.id}.txt"'
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:68]: What does `!Sub` do?

File: content/blog/command-provider-ga/index.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+- Running a command locally after creating a resource, to register it with an external service
+- Running a command locally before deleting a resource, to deregister it with an external service
+- Running a command on a remote host immediately after creating it
+- Copying a file to a remote host after creating it (potentially as a script to be executed afterwards)
+- As a simple alternative to some use cases for Dynamic Providers, by running appropriate commands in the different stages of the Pulumi life cycle.
+
+Here’s a simple example of running commands when another resource changes. In this case, a file is created when a pet name is created or changed, and the file is deleted when the pet name is removed. In a real-world scenario, the command could register/deregister the resource with an external API.
+
+{{< chooser language "typescript,go,python,csharp,yaml" >}}
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:55]: Should we generate a java example?

File: content/blog/command-provider-ga/index.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+});
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+Rather than simply putting a “v1” label on the existing provider, we used this opportunity to enhance it with some previously requested capabilities.
+
+- The former CopyFile resource was renamed to CopyToRemote and can now copy whole directories in addition to individual files.
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:181]:
```suggestion
- The former `CopyFile` resource was renamed to `CopyToRemote` and can now copy whole directories in addition to individual files.
```

File: content/blog/command-provider-ga/index.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+- The source of the copy is now a Pulumi [asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/). This provides full interoperability with the Pulumi ecosystem and is the same as in, for instance, AWS Bucket uploads.
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:182]: The last fragment ("is the same as in, for instance,") is pretty hard to parse.
```suggestion
- The source of the copy is now a [Pulumi asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/). This provides full interoperability with the Pulumi ecosystem.
```

File: content/blog/command-provider-ga/index.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+- The use of assets and archives also makes Pulumi run copy operations only if the source has changed.
+- Capturing stdout and stderr of commands can now be switched off, which is useful when they might contain secrets or are very noisy.
+- There are now official examples in the Pulumi examples repository, and examples have also been added to the API documentation in the Pulumi registry.
+- Environment handling for remote commands was fixed and better documented.
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:186]:
```suggestion
- Environment handling for remote commands was fixed and is better documented.
```

File: content/blog/command-provider-ga/index.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
+---
+title: "Announcing the general availability of the Command provider"
+
+date: 2024-06-20T11:02:20+02:00
+
+draft: false
```
Comment (iwahbe) [inline content/blog/command-provider-ga/index.md:6]: What does `draft` do?

Comment (thomas11) [reply content/blog/command-provider-ga/index.md:6]: > # The draft setting determines whether a post is published. Set it to true if
> # you want to be able to merge the post without publishing it.

from archetypes/blog-post/index.md

I'm not 100% sure how to interpret this. I think we'll merge this PR once it's ready to publish, therefore I set it to true.

File: content/blog/command-provider-ga/index.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+# https://twitter.com/PulumiCorp/status/1763265391042654623
+# https://twitter.com/PulumiCorp/status/1762900472489185492
+# https://twitter.com/PulumiCorp/status/1755637618631405655
+
+social:
+    twitter:
+    linkedin:
+---
+
+Today, we’re happy to announce version v1.0.0 of the Pulumi Command provider. This release marks the provider’s official transition from preview status to general availability (GA).
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:38]:
```suggestion
Today, we’re happy to announce the 1.0 release of the [Pulumi Command](https://www.pulumi.com/registry/packages/command/) provider. This release marks the provider’s official transition from preview status to general availability (GA).
```

File: content/blog/command-provider-ga/index.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+Since its first preview release in late 2021, many users found the flexibility of the Command package valuable. In fact, it’s currently our tenth-most popular provider, so it's time to make things official!
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:39]:
```suggestion
Since its first preview release in late 2021, thousands of Pulumi users incorporated the Command package into their cloud infrastructure projects to help manage local and remote command execution and filesystem operations. In fact, the Command provider is already our tenth-most popular provider, so it's time to make things official!
```

FWIW - I see 5256 users who have used it in Pulumi Cloud, and actually that it is the 9th most popular :-).

File: content/blog/command-provider-ga/index.md [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+
+<!--more-->
+
+The Pulumi Command Provider enables you to execute commands and scripts either locally or remotely as part of the Pulumi resource model, enabling stateful command execution. It also has convenient support for copying assets via SSH. Being able to run arbitrary commands provides great flexibility.
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:43]:
```suggestion
The Pulumi Command Provider enables you to execute commands and scripts either locally or remotely as part of the Pulumi resource model, enabling stateful command execution. It also has convenient support for copying assets via SSH. Being able to run arbitrary commands opens up a wide variety of scenarios and integrations with other tools and systems.
```

File: content/blog/command-provider-ga/index.md [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+- Running a command locally after creating a resource, to register it with an external service
+- Running a command locally before deleting a resource, to deregister it with an external service
+- Running a command on a remote host immediately after creating it
+- Copying a file to a remote host after creating it (potentially as a script to be executed afterwards)
+- As a simple alternative to some use cases for Dynamic Providers, by running appropriate commands in the different stages of the Pulumi life cycle.
+
+Here’s a simple example of running commands when another resource changes. In this case, a file is created when a pet name is created or changed, and the file is deleted when the pet name is removed. In a real-world scenario, the command could register/deregister the resource with an external API.
+
+{{< chooser language "typescript,go,python,csharp,yaml" >}}
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:55]: I think we could use a more "exciting" example.  A few things:
1. Could we make one of the examples at https://www.pulumi.com/registry/packages/command/ simple enough to include here?
2. I think using `RandomPet` may be distracting - it's not a "normal" infrastructure thing, and not really core to what we're trying to talk about here.
3. If none of the existing examples above work, would still be good to describe a "real" scenario here of some sort

File: content/blog/command-provider-ga/index.md [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        },
+    });
+});
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+Rather than simply putting a “v1” label on the existing provider, we used this opportunity to enhance it with some previously requested capabilities.
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:203]: I think we need to say more about what 1.0 means:
1. A stable API for the 1.x series
2. A complete set of capabilities for the core user needs for the Command provider

And then talk about a few of the recent things we've done to finalize (2) over the last few months ahead of the 1.0 release.

File: content/blog/command-provider-ga/index.md [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+- The former `CopyFile` resource was renamed to `CopyToRemote` and can now copy whole directories in addition to individual files.
+- The source of the copy is now a [Pulumi asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/). This provides full interoperability with the Pulumi ecosystem.
+- The use of assets and archives also makes Pulumi run copy operations only if the source has changed.
+- Capturing stdout and stderr of commands can now be switched off, which is useful when they might contain secrets or are very noisy.
+- There are now official examples in the Pulumi examples repository, and examples have also been added to the API documentation in the Pulumi registry.
+- Environment handling for remote commands was fixed and is better documented.
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:210]: I would change the order around:

```
- There are now official examples in the Pulumi examples repository, and examples have also been added to the API documentation in the Pulumi registry.
- Capturing stdout and stderr of commands can now be switched off, which is useful when they might contain secrets or are very noisy.
- The former `CopyFile` resource was renamed to `CopyToRemote` and can now copy whole directories in addition to individual files, as well as using [Pulumi asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/) as input.
- Environment handling for remote commands was fixed and is better documented.
```

Can we also add links to more details on each of these?

File: content/blog/command-provider-ga/index.md [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+
+Here’s an example of copying a directory to a remote host.
+**TODO: Insert remote copy directory example here**
+
+{{< chooser language "typescript,go,python,csharp,yaml" >}}
+
+{{< /chooser >}}
+
+**TODO**: Insert section on breaking changes and migrating here. Depends on whether we decide to keep the existing `CopyFile` resource with a deprecation notice.
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:219]: My vote is to keep the existing `CopyFile` with a deprecation notice. And to not have any "migration guide" be necessary here.

File: content/blog/command-provider-ga/index.md [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+
+{{< /chooser >}}
+
+Rather than simply putting a “v1” label on the existing provider, we used this opportunity to enhance it with some previously requested capabilities.
+
+- The former `CopyFile` resource was renamed to `CopyToRemote` and can now copy whole directories in addition to individual files.
+- The source of the copy is now a [Pulumi asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/). This provides full interoperability with the Pulumi ecosystem.
+- The use of assets and archives also makes Pulumi run copy operations only if the source has changed.
+- Capturing stdout and stderr of commands can now be switched off, which is useful when they might contain secrets or are very noisy.
+- There are now official examples in the Pulumi examples repository, and examples have also been added to the API documentation in the Pulumi registry.
```
Comment (lukehoban) [inline content/blog/command-provider-ga/index.md:209]: BTW - I don't see these yet - and also don't see the examples in https://www.pulumi.com/registry/packages/command/ in all languages yet.  Will those be updated as part of the launch?

**Review by lukehoban — COMMENTED** [review]
A few suggestions - but overall looks great - looking forward to the launch~

File: content/blog/command-provider-ga/index.md (hunk 11 above)
Comment (thomas11) [reply content/blog/command-provider-ga/index.md]: replaced with a working Lambda example

File: content/blog/command-provider-ga/index.md (hunk 15 above)
Comment (thomas11) [reply content/blog/command-provider-ga/index.md]: https://github.com/pulumi/pulumi-command/pull/477

File: content/blog/command-provider-ga/index.md [hunk 16]
```diff
@@ -0,0 +1,750 @@
...
+
+- The API documentation in the Pulumi registry has [examples in all Pulumi languages](https://github.com/pulumi/pulumi-command/issues/196) and is expanded.
+- Capturing stdout and stderr of commands can now [be switched off](https://github.com/pulumi/pulumi-command/pull/451), which is useful when they might contain secrets or are very noisy.
+- Environment handling for remote commands was [fixed and is better documented](https://github.com/pulumi/pulumi-command/pull/395).
+- The `CopyFile` resource is [superseded](https://github.com/pulumi/pulumi-command/pull/423) by the new `CopyToRemote` resource. It can copy whole directories in addition to individual files. The source of the copy is now a [Pulumi asset or archive](https://www.pulumi.com/docs/concepts/assets-archives/) which provides full interoperability with the Pulumi ecosystem. The use of assets and archives also makes Pulumi run copy operations only if the source has changed. For an easy transition, the previous `CopyFile` resource will remain available for a while with a deprecation notice.
```
Comment (mjeffryes) [inline content/blog/command-provider-ga/index.md:396]: rather than saying "for a while" why don't we commit to keeping it around until the next major version of command? (We know this will hold until then since we don't like to make a breaking change like that in a minor and is sounds a little more reassuring)

File: content/blog/command-provider-ga/index.md [hunk 17]
```diff
@@ -0,0 +1,750 @@
...
+
+{{% /choosable %}}
+
+## What's new
+
+The 1.0 release of the Command provider marks a stable API for the 1.x series. Rather than simply putting a “v1” label on the existing provider, we also used this opportunity to enhance it with some previously requested capabilities to provide a more complete set of capabilities.
+
+- The API documentation in the Pulumi registry has [examples in all Pulumi languages](https://github.com/pulumi/pulumi-command/issues/196) and is expanded.
+- Capturing stdout and stderr of commands can now [be switched off](https://github.com/pulumi/pulumi-command/pull/451), which is useful when they might contain secrets or are very noisy.
+- Environment handling for remote commands was [fixed and is better documented](https://github.com/pulumi/pulumi-command/pull/395).
```
Comment (mjeffryes) [inline content/blog/command-provider-ga/index.md:395]: "was fixed" is a little vague "now returns better errors" is the meat of the fix right?

File: content/blog/command-provider-ga/index.md [hunk 18]
```diff
@@ -0,0 +1,750 @@
...
+  output: ${invokeCommand.stdout}
+```
+
+{{% /choosable %}}
+
+## What's new
+
+The 1.0 release of the Command provider marks a stable API for the 1.x series. Rather than simply putting a “v1” label on the existing provider, we also used this opportunity to enhance it with some previously requested capabilities to provide a more complete set of capabilities.
+
+- The API documentation in the Pulumi registry has [examples in all Pulumi languages](https://github.com/pulumi/pulumi-command/issues/196) and is expanded.
```
Comment (mjeffryes) [inline content/blog/command-provider-ga/index.md:393]: FWIW this reads better to me:

"Expanded API documentation in the Pulumi registry and added [examples in all Pulumi languages](https://github.com/pulumi/pulumi-command/issues/196)"

**Review by mjeffryes — APPROVED** [review]
A couple nits on wording, but overall this looks pretty good to me.
